repo/blob/gdrive/gdrive_storage.go

```diff
@@ -0,0 +1,460 @@
+		file, err = gdrive.QueryForBlob(ctx, b, listMetadataFields)
+	}
+	if err != nil {
+		return blob.Metadata{}, errors.Wrap(err, "get blob metadata")
+	}
+
+	bm, err := parseBlobMetadata(file, b)
+	if err != nil {
+		return blob.Metadata{}, err
+	}
+
+	return bm, nil
+}
+
+func (gdrive *gdriveStorage) PutBlob(ctx context.Context, b blob.ID, data blob.Bytes, opts blob.PutOptions) error {
+	switch {
+	case opts.HasRetentionOptions():
+		return errors.Wrap(blob.ErrUnsupportedPutBlobOption, "blob-retention")
+	}
+
+	fileId, err := gdrive.GetFileId(ctx, b)
+	existingFile := true
+	if errors.Is(err, blob.ErrBlobNotFound) {
+		existingFile = false
+	} else if err != nil {
+		return err
+	}
+	if existingFile && opts.DoNotRecreate {
+		return blob.ErrBlobAlreadyExists
+	}
+
+	var file *drive.File
+	mtime := ""
+
+	if !opts.SetModTime.IsZero() {
+		mtime = opts.SetModTime.Format(time.RFC3339)
+	}
+
+	if !existingFile {
+		file, err = gdrive.client.Files.Create(&drive.File{
+			Name:         toFileName(b),
+			Parents:      []string{gdrive.folderId},
+			MimeType:     gdriveMimeType,
+			ModifiedTime: mtime,
+		}).Media(data.Reader(),
+			googleapi.ChunkSize(uploadChunkSize),
+			googleapi.ContentType(uploadContentType),
+		).Context(ctx).Do()
+	} else {
+		file, err = gdrive.client.Files.Update(fileId, &drive.File{
+			ModifiedTime: mtime,
+		}).Media(
+			data.Reader(),
+			googleapi.ChunkSize(uploadChunkSize),
+			googleapi.ContentType(uploadContentType),
+		).Context(ctx).Do()
+	}
+	if err != nil {
+		return translateError(err)
+	}
+
+	gdrive.fileIdCache[b] = file.Id
+	if opts.GetModTime != nil {
+		if mtime, err := parseModifiedTime(file); err == nil {
+			*opts.GetModTime = mtime
+		}
+	}
+
+	return nil
+}
+
+func (gdrive *gdriveStorage) DeleteBlob(ctx context.Context, b blob.ID) error {
+	fileId, err := gdrive.GetFileId(ctx, b)
+	if errors.Is(err, blob.ErrBlobNotFound) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	err = gdrive.client.Files.Delete(fileId).Context(ctx).Do()
+	return translateError(err)
+}
+
+func (gdrive *gdriveStorage) ListBlobs(ctx context.Context, prefix blob.ID, callback func(blob.Metadata) error) error {
+	// Tracks blob matches in cache but not returned by API.
+	unvisitedIds := make(map[blob.ID]bool)
+
+	consumer := func(files *drive.FileList) error {
+		for _, file := range files.Files {
+			blobId := toBlobId(file.Name)
+			gdrive.fileIdCache[blobId] = file.Id
+
+			if !matchesPrefix(blobId, prefix) {
+				return nil
+			}
+
+			// Mark blob as visited.
+			delete(unvisitedIds, blobId)
+
+			bm, err := parseBlobMetadata(file, blobId)
+			if err != nil {
+				return err
+			}
+
+			if err := callback(bm); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	query := fmt.Sprintf("'%s' in parents and mimeType = '%s'", gdrive.folderId, gdriveMimeType)
+	if prefix != "" {
+		query = fmt.Sprintf("'%s' in parents and name contains '%s' and mimeType = '%s'", gdrive.folderId, capPrefix(prefix), gdriveMimeType)
```

Comment:
should it be `contains`? is there 'startsWith' or something like that?

Reply:
So according to the drive api `contains` acts like `startswith` for file names. I added a comment here to clarify this.